### PR TITLE
✨Fill provider types with values in operator

### PR DESCRIPTION
**Pull request by alexander-demichev**
**What this PR does / why we need it**:

Fill provider types with values in operator according to proposal https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20201020-capi-provider-operator.md#new-api-types

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

**Review by fabriziopandini — COMMENTED** [review]
I don't have strong opinions here, but the original agreement was to proceed "use case by use case", adding API fields and controller logic incrementally.
Are we changing approach?

**Comment by alexander-demichev** [comment]
I'm fine with both approaches, happy to close this PR and add fields incrementally.

**Comment by fabriziopandini** [comment]
You are driving here, the last call is on you 😉 
I just need some context so I can try to provide more meaningful reviews comments

**Review by MarcelMue — COMMENTED** [review]
I am fine with taking this approach :+1:

/lgtm

File: exp/operator/api/v1alpha4/provider_types.go [hunk 1]
```diff
@@ -16,11 +16,180 @@ limitations under the License.
...
+	// List of environment variables to set in the container.
+	// +optional
+	Env []corev1.EnvVar `json:"env,omitempty"`
+
+	// Compute resources required by this container.
+	// +optional
+	Resources *corev1.ResourceRequirements `json:"resources,omitempty"`
+}
+
+// ImageMeta allows to customize the image used
```
Comment (MarcelMue) [inline exp/operator/api/v1alpha4/provider_types.go:150]: Looks like the linter is not happy about these not ending with a `.`

**Comment by alexander-demichev** [comment]
@MarcelMue @fabriziopandini Can you review again? I fixed the linter issues.

File: exp/operator/api/v1alpha4/provider_types.go [hunk 2]
```diff
@@ -16,11 +16,180 @@ limitations under the License.
...
+	Name *string `json:"name,omitempty"`
+
+	// Tag allows to specify a tag for the image.
+	// +optional
+	Tag *string `json:"tag,omitempty"`
+}
+
+// FetchConfiguration determines the way to fetch the components and metadata for the provider.
+type FetchConfiguration struct {
+	// URL to be used for fetching the provider’s components and metadata from a remote Github repository.
```
Comment (MarcelMue) [inline exp/operator/api/v1alpha4/provider_types.go:167]:
```suggestion
	// URL to be used for fetching the providers components and metadata from a remote Github repository.
```

File: exp/operator/api/v1alpha4/provider_types.go [hunk 3]
```diff
@@ -16,11 +16,180 @@ limitations under the License.
...
+	// For example, https://github.com/{owner}/{repository}/releases
```
Comment (MarcelMue) [inline exp/operator/api/v1alpha4/provider_types.go:168]:
```suggestion
	// For example: https://github.com/{owner}/{repository}/releases
```

File: exp/operator/api/v1alpha4/provider_types.go [hunk 4]
```diff
@@ -16,11 +16,180 @@ limitations under the License.
...
+}
+
+// ContainerSpec defines the properties available to override for each
+// container in a provider deployment such as Image and Args to the container’s
+// entrypoint.
+type ContainerSpec struct {
+	// Name of the container. Cannot be updated.
+	Name string `json:"name"`
+
+	// Container Image Name
```
Comment (MarcelMue) [inline exp/operator/api/v1alpha4/provider_types.go:128]:
```suggestion
	// Container Image Name.
```

File: exp/operator/api/v1alpha4/provider_types.go [hunk 5]
```diff
@@ -16,11 +16,180 @@ limitations under the License.
...
+	Resources *corev1.ResourceRequirements `json:"resources,omitempty"`
+}
+
+// ImageMeta allows to customize the image used.
+type ImageMeta struct {
+	// Repository sets the container registry to pull images from.
+	// +optional
+	Repository *string `json:"repository,omitempty"`
+
+	// Name allows to specify a name for the image.
```
Comment (MarcelMue) [inline exp/operator/api/v1alpha4/provider_types.go:156]:
```suggestion
	// Name of the image.
```

File: exp/operator/api/v1alpha4/provider_types.go [hunk 6]
```diff
@@ -16,11 +16,180 @@ limitations under the License.
...
+	// +optional
+	Name *string `json:"name,omitempty"`
+
+	// Tag allows to specify a tag for the image.
```
Comment (MarcelMue) [inline exp/operator/api/v1alpha4/provider_types.go:160]:
```suggestion
	// Tag of the image.
```

File: exp/operator/api/v1alpha4/provider_types.go [hunk 7]
```diff
@@ -16,11 +16,180 @@ limitations under the License.
...
+
+	// If specified, the pod's tolerations.
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+
+	// If specified, the pod's scheduling constraints
+	// +optional
+	Affinity *corev1.Affinity `json:"affinity,omitempty"`
+
+	// List of containers specified in the Deployment
```
Comment (MarcelMue) [inline exp/operator/api/v1alpha4/provider_types.go:116]:
```suggestion
	// List of containers specified in the Deployment.
```

File: exp/operator/api/v1alpha4/provider_types.go [hunk 8]
```diff
@@ -16,11 +16,180 @@ limitations under the License.
...
+	// Selector which must match a node's labels for the pod to be scheduled on that node.
+	// More info: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+
+	// If specified, the pod's tolerations.
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+
+	// If specified, the pod's scheduling constraints
```
Comment (MarcelMue) [inline exp/operator/api/v1alpha4/provider_types.go:112]:
```suggestion
	// Scheduling constraints of the pod.
```

File: exp/operator/api/v1alpha4/provider_types.go [hunk 9]
```diff
@@ -16,11 +16,180 @@ limitations under the License.
...
+	// +optional
+	Replicas *int `json:"replicas,omitempty"`
+
+	// NodeSelector is a selector which must be true for the pod to fit on a node.
+	// Selector which must match a node's labels for the pod to be scheduled on that node.
+	// More info: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+
+	// If specified, the pod's tolerations.
```
Comment (MarcelMue) [inline exp/operator/api/v1alpha4/provider_types.go:108]:
```suggestion
	// Tolerations of the pod.
```

File: exp/operator/api/v1alpha4/provider_types.go [hunk 10]
```diff
@@ -16,11 +16,180 @@ limitations under the License.
...
+	// which can be run. Defaults to 10.
+	// +optional
+	MaxConcurrentReconciles *int `json:"maxConcurrentReconciles,omitempty"`
+
+	// Verbosity set the logs verbosity. Defaults to 1.
+	// Controller Manager flag is --verbosity.
+	// +optional
+	Verbosity int `json:"verbosity,omitempty"`
+
+	// Debug, if set, will override a set of fields with opinionated values for
```
Comment (MarcelMue) [inline exp/operator/api/v1alpha4/provider_types.go:85]:
```suggestion
	// Debug will override a set of fields with opinionated values for
```

**Review by MarcelMue — COMMENTED** [review]
Super quick flyover to check godoc for consistency :)

**Comment by MarcelMue** [comment]
I am not certain why `pull-cluster-api-apidiff-main` is failing - is it safe to ignore?

**Comment by JoelSpeed** [comment]
Since we are still in the experimental phase and the API diff is failing on the CRDs for the operator, it's fine to ignore for now

**Review by MarcelMue — COMMENTED** [review]
We can iterate.

/lgtm

File: exp/operator/api/v1alpha4/bootstrapprovider_types.go [hunk 11]
```diff
@@ -20,19 +20,19 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// BootstrapProviderSpec defines the desired state of BootstrapProvider
+// BootstrapProviderSpec defines the desired state of BootstrapProvider.
```
Comment (fabriziopandini) [inline exp/operator/api/v1alpha4/bootstrapprovider_types.go:23]: If I look at the proposal there is no `BootstrapProviderSpec`, but we are using `ProviderSpec` directly inside the the `BootstrapProvider`.
Is there a reason why you took a different approach here?

File: exp/operator/api/v1alpha4/bootstrapprovider_types.go [hunk 12]
```diff
@@ -20,19 +20,19 @@ import (
...
 type BootstrapProviderSpec struct {
 	ProviderSpec `json:",inline"`
 }
 
-// BootstrapProviderStatus defines the observed state of BootstrapProvider
+// BootstrapProviderStatus defines the observed state of BootstrapProvider.
```
Comment (fabriziopandini) [inline exp/operator/api/v1alpha4/bootstrapprovider_types.go:28]: Same as above (is there a reason why we are deviating from the proposal?)

File: exp/operator/api/v1alpha4/bootstrapprovider_types.go (hunk 11 above)
Comment (fabriziopandini) [reply exp/operator/api/v1alpha4/bootstrapprovider_types.go:23]: This applies to other providers type as well

File: exp/operator/api/v1alpha4/bootstrapprovider_types.go (hunk 12 above)
Comment (fabriziopandini) [reply exp/operator/api/v1alpha4/bootstrapprovider_types.go:28]: This applies to other providers type as well

File: exp/operator/api/v1alpha4/groupversion_info.go [hunk 13]
```diff
@@ -25,10 +25,10 @@ import (
 )
 
 var (
-	// GroupVersion is group version used to register these objects
+	// GroupVersion is group version used to register these objects.
```
Comment (fabriziopandini) [inline exp/operator/api/v1alpha4/groupversion_info.go:28]: Question: if I look at the examples in the proposal, we are using v1alpha1 for this api group. is there  a reason to jump to v1alpha4?

File: exp/operator/api/v1alpha4/provider_types.go [hunk 14]
```diff
@@ -16,11 +16,180 @@ limitations under the License.
...
+type ProviderSpec struct {
+	// Version indicates the provider version.
+	// +optional
+	Version *string `json:"version,omitempty"`
 
-// ProProviderStatus defines the observed state of the Provider.
-type ProviderStatus struct{}
+	// Manager defines the properties that can be enabled on the controller manager for the provider.
+	// +optional
+	Manager ManagerSpec `json:"manager,omitempty"`
```
Comment (fabriziopandini) [inline exp/operator/api/v1alpha4/provider_types.go:34]: Should this be a pointer like we are doing for DeploymentSpec?

File: exp/operator/api/v1alpha4/provider_types.go [hunk 15]
```diff
@@ -16,11 +16,180 @@ limitations under the License.
...
+
+	// Deployment defines the properties that can be enabled on the deployment for the provider.
+	// +optional
+	Deployment *DeploymentSpec `json:"deployment,omitempty"`
+
+	// SecretName is the name of the Secret providing the configuration
```
Comment (fabriziopandini) [inline exp/operator/api/v1alpha4/provider_types.go:40]: NOTE: We should double check this with the effort for multi-tenancy/credential management.
However this is fine for now.

File: exp/operator/api/v1alpha4/groupversion_info.go (hunk 13 above)
Comment (alexander-demichev) [reply exp/operator/api/v1alpha4/groupversion_info.go:28]: Hm, I missed that somehow. Should we stick with v1alpha4 so we can avoid confusion in the future?

File: exp/operator/api/v1alpha4/bootstrapprovider_types.go (hunk 11 above)
Comment (alexander-demichev) [reply exp/operator/api/v1alpha4/bootstrapprovider_types.go:23]: There is no reason, it came from the initial kubebuilder scaffolding. Thank you for noticing it, I'll fix it.

File: exp/operator/api/v1alpha4/groupversion_info.go (hunk 13 above)
Comment (fabriziopandini) [reply exp/operator/api/v1alpha4/groupversion_info.go:28]: My major concern about linking this new API to the other APIs in CAPI is the fact that we will get to v1beta1 pretty quickly without having time to gather enough user feedback.
@vincepri @detiber @CecileRobertMichon opinions?

Comment (vincepri) [reply exp/operator/api/v1alpha4/groupversion_info.go:28]: We can start from v1alpha1 for the operator

File: exp/operator/api/v1alpha4/provider_types.go (hunk 15 above)
Comment (vincepri) [reply exp/operator/api/v1alpha4/provider_types.go:40]: Could we add a TODO in the code?

**Comment by vincepri** [comment]
Hey folks, while this is great progress, I'd prefer to use the approach that we've been using so far, which is not to merge anything that's not also implemented and functional. 

If these fields are not wired up today, and not tested, there might be discrepancies later on when the implementation takes place. Instead, we should merge these once at least an initial functional implementation is done.

The initial PRs can be large, although we can have multiple commits in a single PR and either squash at the end, or keep them separate if they're split in logical groups

**Comment by fabriziopandini** [comment]
Then we can eventually revert to the initial approach and proceed "use case by use case", adding API fields and controller logic incrementally.

**Comment by alexander-demichev** [comment]
Closing this PR, new API changes will be added gradually

**Pull request closed without merging**